src/mem3/README_reshard.md

```diff
@@ -0,0 +1,88 @@
+
+Another interesting aspect is how the `split_state` transitions happen in `mem3_reshard_job`. What follows is an examination of that. A job starts running in the `new` state or from a previously checkpointed state. In the later case, the job goes through some recovery logic (see `?STATE_RESTART` macro in `mem3_reshard_job`) where it tries to resume its work from where it left of. It means, for example, if it was in the `initial_copy` state and was interrupted it might have to reset the target files and copy everything again. After recovery, the state execution logic is driven by `run(#job{})` which ends up calling `?MODULE:State(#job{})` state specific functions for each state.
+
+In `mem3_reshard_job:switch_state/2` job's history is updated, any current `state_info` is cleared, job state is switched in the `#job{}` record and then `mem3_reshard` asked to checkpoint this state. That notificaiton is done via gen_server cast message. After that is send the job process sits and waits. In the meantime `mem3_reshard` manager checkpoints, which mean it updates both its ets table with the new `#job{}` record and also persists the state with the `mem3_reshard_store`, and finally it notifies the job process that checkpointing is done by calling `mem3_reshard_job:checkpoint_done/1`.
+
+`mem3_reshard_job:checkpoint_done/1` function call then sends a `do_state` cast to the job's gen_server which proceeds to execute that particular state.
+
+Most states in `mem3_reshard_job` try not to block the gen_server process and instead launch worker processes to perform long running operations. It is usually just one worker process but it could be multiple as well. After that it mostly waits for the workers to finish and inspects their exit signal. If they all exit normal (see `mem_reshard:worker_exited/2`) it calls `switch_state/2` again with the next state and the whole process repeats until completion. Upon completion, the `mem3_reshard_job` gen_server exits normally.
+
+
+Individual Modules Description
+------------------------------
+
+These are mostly random notes about various modules involved in resharding. Most, but not all, are in the mem3 application.
+
+* `mem3_reshard`: Main API entry point and the job manager.
+
+* `mem3_reshard_job` : Individual job logic.
+
+* `mem3_reshard_dbdoc` : Responsible for updating shard doc in the `_db`'s database. Besides just having a bunch of utility function there is a gen_server spawned which is used to update shard documents in a cluster in such a way as to minimize the risk of conflicts. That is accomplished by having each shard updater calling only one such updater for the whole cluster. This coordinator is picked by sorting the list of all the live mem3 nodes and picking the first one in the list.
```

Comment:
This sounds similar to the replicator scheduler; is it worth mentioning it's a similar approach?

What happens with an unstable cluster, nodes coming on and offline -- could a job jump between nodes and possibly get started multiple times (racey)?

Reply:
There are a few checks before the update starts the node performing the update is still the coordinator. But we don't have a consistent system so there is a small chance of a race. In that case there would be a conflict raised and job would fail, but the shard map should stay as before the split.